File: problem-solving/_lc-153-medium-find-minimum-in-rotated-sorted-array/1-binary-search.go
/*
https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

Complexity
----------
Time:
  Best case: O(logn)
  Worst case: O(n)
Space: O(1)
*/

func findMin(nums []int) int {
    left, right := 0, len(nums) - 1
    for left <= right {
        mid := (left + right) / 2
        if mid + 1 < len(nums) && nums[mid] > nums[mid + 1] {
            return nums[mid + 1]
        } else if nums[left] <= nums[mid] {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return nums[0]
}

File: problem-solving/lc-154-hard-find-minimum-in-rotated-sorted-array-ii/1-binary-search.py
"""
https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/

Complexity
----------
Time:
  Best Case: O(logn)
  Worst Case: O(n)
Space: O(1)
"""

class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            elif nums[mid] < nums[right]:
                right = mid
            else:
                right -= 1
        return nums[left]
